python/kastore/store.py

```diff
@@ -262,23 +260,18 @@ def __init__(self, filename, read_all=False, key_encoding="utf8"):
         self.filename = filename
         self.key_encoding = key_encoding
         self._descriptor_map = None
-        # Make sure _file and _buffer are defined so that we can access them in
-        # the destructor below.
-        self._file = None
-        self._buffer = None
+        self._array = {}
         self._read_all = read_all
-        self._file = open(self.filename, "rb")
-        self._file_size = os.fstat(self._file.fileno()).st_size
-        if self._file_size == 0:
-            raise exceptions.FileFormatError("Empty file")
-        if self._read_all:
-            self._buffer = self._file.read()
-        else:
-            self._buffer = mmap.mmap(self._file.fileno(), 0, access=mmap.ACCESS_READ)
+        self._file = None
```

Comment:
Well, yes. But not entirely. So it needs a rambling comment to explain... :)

The `self._file = open()` call can fail (e.g. if the file doesn't exist), which raises an exception. If an exception is raised, `self._file` is not assigned to. But then, there's a check for if `self._file is not None` in `Store.close()` that gets called from `__exit__` or `__del__`, and thus `self` must have a `_file` attributed, or another exception occurs. I guess there are other ways to deal with this---I can instead put a check for `hasattr(self, "_file")` in `Store.close()` if you prefer that?

Reply:
Ahh! No, setting ``self._file = None`` is the right the thing to do, just it's needs a comment like `` open can fail, and this attribute will be undefined otherwise``, or some busybody will "fix" this later.